File: platform/workspace/rules.py
"""Validated, non-executable test contracts approved by the designer."""
from __future__ import annotations

import hashlib
import json
import re

TARGET = re.compile(r"[A-Za-z][A-Za-z0-9_-]{0,79}\Z")
TEXT_ACTIONS = {"fill", "select", "expectText", "expectValue", "expectStyle"}
BOOL_ACTIONS = {"check", "expectVisible", "expectEnabled", "expectChecked"}
EXPECT_ACTIONS = {"expectText", "expectValue", "expectVisible", "expectEnabled", "expectChecked", "expectStyle"}
ACTIONS = TEXT_ACTIONS | BOOL_ACTIONS | {"click", "press"}
KEYS = {"Enter", "Tab", "Escape", "ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight", "Space"}
STYLE_PROPERTIES = {"color", "backgroundColor", "fontSize", "fontWeight", "fontFamily", "borderRadius",
                    "padding", "margin", "gap", "minHeight", "height", "width", "borderColor", "borderWidth", "display"}
CRITERIA_IDS = ("projectId", "productId", "guidelineId", "guidelineAssetId")
CRITERIA_HASHES = ("catalogHash", "ontologyHash")


def _text(value, label: str, maximum: int, *, empty: bool = False) -> str:
    if not isinstance(value, str) or len(value) > maximum or (not empty and not value.strip()):
        raise ValueError(f"{label}: 1~{maximum}자의 텍스트가 필요합니다.")
    return value


def _integer(value, label: str, low: int, high: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not low <= value <= high:
        raise ValueError(f"{label}: {low}~{high} 범위의 정수가 필요합니다.")
    return value


def validate_contract(data: dict, asset_texts: dict[str, str] | None = None) -> dict:
    if not isinstance(data, dict):
        raise ValueError("동작 규칙은 객체여야 합니다.")
    if type(data.get("schemaVersion", 1)) is not int or data.get("schemaVersion", 1) != 1:
        raise ValueError("지원하지 않는 동작 규칙 버전입니다.")
    asset_ids = data.get("assetIds", [])
    if (not isinstance(asset_ids, list) or len(asset_ids) > 20
            or any(not isinstance(x, str) or not x or len(x) > 128 for x in asset_ids)
            or len(set(asset_ids)) != len(asset_ids)):
        raise ValueError("입력 자산은 중복 없이 최대 20개까지 선택하세요.")
    viewport = data.get("viewport", {"width": 390, "height": 844})
    if not isinstance(viewport, dict):
        raise ValueError("화면 크기가 올바르지 않습니다.")
    normalized = {
        "schemaVersion": 1,
        "title": _text(data.get("title", "동작 규칙"), "규칙 이름", 180),
        "brief": _text(data.get("brief", ""), "화면 설명", 4000, empty=True),
        "assetIds": list(asset_ids),
        "viewport": {
            "width": _integer(viewport.get("width"), "화면 너비", 320, 1920),
            "height": _integer(viewport.get("height"), "화면 높이", 480, 2160),
        },
        "rules": [],
        "unresolved": [],
        "bindings": {},
    }
    for name in CRITERIA_IDS:
        if name in data:
            value = data[name]
            if not isinstance(value, str) or not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_-]{0,127}", value):
                raise ValueError(f"{name}: 기준 버전 식별자가 올바르지 않습니다.")
            normalized[name] = value
    for name in CRITERIA_HASHES:
        if name in data:
            if not isinstance(data[name], str) or not re.fullmatch(r"[a-f0-9]{64}", data[name]):
                raise ValueError(f"{name}: 기준 코드/온톨로지 해시가 올바르지 않습니다.")
            normalized[name] = data[name]
    if any(name in normalized for name in CRITERIA_IDS):
        if not all(name in normalized for name in (*CRITERIA_IDS, *CRITERIA_HASHES)):
            raise ValueError("프로젝트·상품·가이드·컴포넌트 기준을 함께 고정해야 합니다.")
        if normalized["guidelineAssetId"] not in asset_ids:
            raise ValueError("확정된 기획 가이드가 입력 자산에 포함되어야 합니다.")
    bindings = data.get("bindings", {})
    if not isinstance(bindings, dict) or len(bindings) > 100:
        raise ValueError("원본 요소 연결은 최대 100개까지 지정할 수 있습니다.")
    for target, selector in bindings.items():
        if (not isinstance(target, str) or not TARGET.fullmatch(target) or not isinstance(selector, str)
                or not selector.strip() or len(selector) > 300 or ">>" in selector
                or any(char in selector for char in "\r\n\x00") or selector.startswith(("xpath=", "text=", "javascript:"))):
            raise ValueError("원본 요소는 안전한 CSS 선택자로 연결하세요.")
        normalized["bindings"][target] = selector
    unresolved = data.get("unresolved", [])
    if not isinstance(unresolved, list) or len(unresolved) > 30:
        raise ValueError("미정의 요구사항은 최대 30개까지 기록할 수 있습니다.")
    normalized["unresolved"] = [_text(x, "미정의 요구사항", 1000) for x in unresolved]
    rules = data.get("rules")
    if not isinstance(rules, list) or not 1 <= len(rules) <= 20:
        raise ValueError("검증할 동작 규칙을 1~20개 작성하세요.")
    ids = set()
    for rule in rules:
        if not isinstance(rule, dict):
            raise ValueError("각 동작 규칙은 객체여야 합니다.")
        rid = _text(rule.get("id"), "규칙 ID", 80)
        if rid in ids or not TARGET.fullmatch(rid):
            raise ValueError("규칙 ID는 중복 없는 영문 식별자여야 합니다.")
        ids.add(rid)
        if not isinstance(rule.get("required", True), bool):
            raise ValueError("필수 여부는 참/거짓으로 지정하세요.")
        source = rule.get("source", {"kind": "manual"})
        if not isinstance(source, dict) or source.get("kind") not in ("explicit", "inferred", "manual"):
            raise ValueError("규칙의 출처를 명시/추정/직접 작성으로 구분하세요.")
        clean_source = {"kind": source["kind"]}
        if source.get("assetId"):
            if source["assetId"] not in asset_ids:
                raise ValueError("규칙 출처는 선택한 반입 파일이어야 합니다.")
            clean_source["assetId"] = source["assetId"]
        if source["kind"] == "explicit":
            if not clean_source.get("assetId"):
                raise ValueError("명시된 규칙에는 출처 파일이 필요합니다.")
            quote = _text(source.get("quote"), "원문 근거", 2000)
            if asset_texts is not None:
                original = asset_texts.get(source["assetId"], "")
                # Whitespace normalization only; no fabricated or fuzzy quotes.
                if " ".join(quote.split()) not in " ".join(original.split()):
                    raise ValueError(f"{rid}: 원문 근거가 선택한 파일에 없습니다.")
            clean_source["quote"] = quote
        elif source.get("quote"):
            clean_source["quote"] = _text(source["quote"], "참고 설명", 2000)
        if "page" in source:
            clean_source["page"] = _integer(source["page"], "출처 페이지", 1, 10000)
        steps = rule.get("steps")
        if not isinstance(steps, list) or not 1 <= len(steps) <= 20:
            raise ValueError(f"{rid}: 조작과 확인 단계를 1~20개 지정하세요.")
        clean_steps = []
        for step in steps:
            if not isinstance(step, dict) or step.get("action") not in ACTIONS:
                raise ValueError(f"{rid}: 지원하지 않는 검증 동작입니다.")
            action, target = step["action"], step.get("target")
            if not isinstance(target, str) or not TARGET.fullmatch(target):
                raise ValueError(f"{rid}: 대상은 안전한 data-testid 식별자여야 합니다.")
            clean = {"action": action, "target": target,
                     "targetLabel": _text(step.get("targetLabel", target), "대상 이름", 180)}
            if action in TEXT_ACTIONS:
                clean["value"] = _text(step.get("value"), "입력/기대값", 2000, empty=True)
            elif action in BOOL_ACTIONS:
                if not isinstance(step.get("value"), bool):
                    raise ValueError(f"{rid}: {action} 값은 참/거짓이어야 합니다.")
                clean["value"] = step["value"]
            elif action == "press":
                if step.get("value") not in KEYS:
                    raise ValueError(f"{rid}: 지원하지 않는 키 입력입니다.")
                clean["value"] = step["value"]
            if action == "expectText":
                match = step.get("match", "contains")
                if match not in ("contains", "equals"):
                    raise ValueError("문구 비교는 포함 또는 일치만 지원합니다.")
                clean["match"] = match
                if "normalizeWhitespace" in step:
                    if type(step["normalizeWhitespace"]) is not bool:
                        raise ValueError("공백 정규화 여부는 참/거짓이어야 합니다.")
                    clean["normalizeWhitespace"] = step["normalizeWhitespace"]
                if not clean["value"]:
                    raise ValueError("문구 확인의 기대값을 비워둘 수 없습니다.")
            if action == "expectStyle":
                if step.get("property") not in STYLE_PROPERTIES or not clean["value"].strip():
                    raise ValueError("스타일 검사는 지원하는 속성과 기대값을 지정하세요.")
                clean["property"] = step["property"]
            clean_steps.append(clean)
        if not any(s["action"] in EXPECT_ACTIONS for s in clean_steps):
            raise ValueError(f"{rid}: 실제 결과를 확인하는 단계가 필요합니다.")
        normalized["rules"].append({
            "id": rid, "title": _text(rule.get("title"), "동작 설명", 500),
            "required": rule.get("required", True), "source": clean_source, "steps": clean_steps,
        })
    if not any(r["required"] for r in normalized["rules"]):
        raise ValueError("최소 한 개의 필수 동작 규칙이 필요합니다.")
    if not any(step["action"] in EXPECT_ACTIONS and
               (step["action"] != "expectVisible" or step["value"] is True)
               for rule in normalized["rules"] for step in rule["steps"]):
        raise ValueError("실제로 존재하는 화면 요소를 확인하는 단계를 최소 하나 추가하세요.")
    return normalized


def contract_hash(contract: dict) -> str:
    normalized = validate_contract(contract)
    return hashlib.sha256(json.dumps(normalized, ensure_ascii=False, sort_keys=True,
                                     separators=(",", ":")).encode()).hexdigest()


def report_passes(contract: dict, report: dict, *, visual_required: bool = False) -> bool:
    """A passing boolean alone is never sufficient evidence."""
    if (not isinstance(report, dict) or report.get("passed") is not True
            or report.get("functionalStatus") != "pass"
            or report.get("networkRequests") != [] or report.get("consoleErrors") != []
            or report.get("blockingFindings") != []):
        return False
    accessibility, visual = report.get("accessibility") or {}, report.get("visual") or {}
    if accessibility.get("status") != "pass" or accessibility.get("incomplete"):
        return False
    if visual.get("status") not in (("pass",) if visual_required else ("pass", "not-run")):
        return False
    rules, checks = contract.get("rules", []), report.get("checks")
    if not rules or not isinstance(checks, list) or len(checks) != len(rules):
        return False
    by_id = {}
    for check in checks:
        if not isinstance(check, dict) or check.get("caseId") in by_id or check.get("status") not in ("pass", "fail"):
            return False
        by_id[check.get("caseId")] = check
    for rule in rules:
        check = by_id.get(rule["id"])
        if not check or (rule.get("required", True) and check["status"] != "pass"):
            return False
        if check["status"] == "pass":
            actual = check.get("steps")
            if not isinstance(actual, list) or len(actual) != len(rule["steps"]):
                return False
            for index, (step, evidence) in enumerate(zip(rule["steps"], actual)):
                if (not isinstance(evidence, dict) or evidence.get("index") != index
                        or evidence.get("status") != "pass"
                        or evidence.get("action") != step["action"] or evidence.get("target") != step["target"]
                        or evidence.get("expected") != step.get("value")
                        or (step["action"] == "expectStyle" and evidence.get("property") != step["property"])):
                    return False
    return True

File: platform/web/src/workspace/rules.ts
import type { Action, Asset, EditableContract, Purpose, Round, Rule, Run, Step, StyleProperty } from './types';

export const PURPOSES = {
  prototype: 'HTML 화면 시안', reference: '화면 기준 이미지', component: '컴포넌트·아이콘 자산',
  guide: '업무·스타일 가이드', skill: '작업 지침·스킬', token: '디자인 토큰', archive: '원본 보관',
} as const;
export function suggestedPurpose(name: string): Purpose {
  const extension = name.split('.').pop()?.toLowerCase();
  if (extension === 'html' || extension === 'htm') return 'prototype';
  if (extension === 'css') return 'guide';
  if (['png', 'jpg', 'jpeg'].includes(extension || '')) return 'reference';
  if (extension === 'svg') return 'component';
  if (extension === 'fig') return 'archive';
  if (/^skill\.m(?:d|arkdown)$/i.test(name)) return 'skill';
  return 'guide';
}
export const ACTIONS: Record<Action, string> = {
  fill: '값 입력', click: '누르기', check: '체크 변경', select: '항목 선택', press: '키 누르기',
  expectText: '문구 확인', expectValue: '입력값 확인', expectVisible: '표시 여부 확인',
  expectEnabled: '사용 가능 여부 확인', expectChecked: '체크 여부 확인', expectStyle: '스타일 일치',
};
export const STYLE_PROPERTIES: Record<StyleProperty, string> = {
  color: '글자 색', backgroundColor: '배경 색', fontSize: '글자 크기', fontWeight: '글자 굵기', fontFamily: '글꼴',
  borderRadius: '모서리 둥글기', padding: '안쪽 여백', margin: '바깥 여백', gap: '요소 간격',
  minHeight: '최소 높이', height: '높이', width: '너비', borderColor: '테두리 색', borderWidth: '테두리 두께', display: '표시 방식',
};
export const BOOLEAN_ACTIONS: Action[] = ['check', 'expectVisible', 'expectEnabled', 'expectChecked'];
export const KEYS = ['Enter', 'Tab', 'Escape', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', 'Space'];
export function newStep(action: Action = 'expectVisible'): Step {
  return { action, target: `field_${crypto.randomUUID().replace(/-/g, '').slice(0, 10)}`, targetLabel: '확인할 화면 요소',
    ...(action === 'click' ? {} : { value: BOOLEAN_ACTIONS.includes(action) ? true : action === 'press' ? 'Enter' : '' }),
    ...(action === 'expectText' ? { match: 'contains' as const } : {}),
    ...(action === 'expectStyle' ? { property: 'color' as const } : {}) };
}
export function newRule(): Rule {
  return { id: `R_${crypto.randomUUID().slice(0, 8)}`, title: '화면 요소가 표시된다', required: true, source: { kind: 'manual' }, steps: [newStep()] };
}
export function contractProblems(contract: EditableContract): string[] {
  const problems: string[] = [];
  if (!contract.title.trim()) problems.push('규칙 묶음의 이름을 입력하세요.');
  if (contract.brief.length > 4000) problems.push('만들 화면 설명은 4,000자 이내로 입력하세요.');
  if (!contract.rules.length || contract.rules.length > 20) problems.push('규칙은 1~20개가 필요합니다.');
  if (contract.assetIds.length > 20) problems.push('파일은 최대 20개까지 선택하세요.');
  if (contract.unresolved.some(value => value.trim())) problems.push('확인이 필요한 내용을 모두 해결해야 승인할 수 있습니다.');
  if (!Number.isInteger(contract.viewport.width) || !Number.isInteger(contract.viewport.height) ||
      contract.viewport.width < 320 || contract.viewport.width > 1920 || contract.viewport.height < 480 || contract.viewport.height > 2160) {
    problems.push('화면 너비는 320~1920, 높이는 480~2160 사이의 정수로 입력하세요.');
  }
  if (!contract.rules.some(rule => rule.required)) problems.push('필수 확인 규칙이 최소 한 개 필요합니다.');
  if (contract.bindings !== undefined) {
    if (!contract.bindings || typeof contract.bindings !== 'object' || Array.isArray(contract.bindings) ||
        Object.keys(contract.bindings).length > 100) problems.push('원본 HTML 요소 연결은 최대 100개까지 지정하세요.');
    else for (const [target, selector] of Object.entries(contract.bindings)) {
      if (!/^[A-Za-z][A-Za-z0-9_-]{0,79}$/.test(target) || typeof selector !== 'string' || !selector.trim() ||
          selector.length > 300 || selector.includes('>>') || /[\r\n\0]/.test(selector) || /^(xpath=|text=|javascript:)/.test(selector)) {
        problems.push('고급 설정의 원본 HTML 요소 연결에는 300자 이내의 CSS 선택자를 입력하세요. XPath·실행식·선택자 연결은 사용할 수 없습니다.');
      }
    }
  }
  for (const [index, rule] of contract.rules.entries()) {
    const name = `규칙 ${index + 1}`;
    if (!rule.title.trim()) problems.push(`${name}: 규칙 이름을 입력하세요.`);
    if (!rule.steps.length || rule.steps.length > 20) problems.push(`${name}: 단계는 1~20개가 필요합니다.`);
    if (!rule.steps.some(step => step.action.startsWith('expect'))) problems.push(`${name}: 결과를 확인하는 단계를 추가하세요.`);
    for (const step of rule.steps) {
      if (!(step.action in ACTIONS) || !/^[A-Za-z][A-Za-z0-9_-]{0,79}$/.test(step.target)) problems.push(`${name}: 고급 설정의 요소 식별자를 확인하세요.`);
      if (!step.targetLabel.trim()) problems.push(`${name}: 화면 요소 이름을 입력하세요.`);
      if (BOOLEAN_ACTIONS.includes(step.action) && typeof step.value !== 'boolean') problems.push(`${name}: 여부를 선택하세요.`);
      if (!BOOLEAN_ACTIONS.includes(step.action) && step.action !== 'click' && typeof step.value !== 'string') problems.push(`${name}: 확인할 값을 입력하세요.`);
      if (step.action === 'expectText' && !step.value) problems.push(`${name}: 확인할 문구를 입력하세요.`);
      if (step.action === 'expectText' && step.normalizeWhitespace !== undefined && typeof step.normalizeWhitespace !== 'boolean')
        problems.push(`${name}: 공백 정규화 여부는 예 또는 아니요로 선택하세요.`);
      if (step.action === 'expectStyle' && (!step.property || !Object.hasOwn(STYLE_PROPERTIES, step.property) ||
          typeof step.value !== 'string' || !step.value.trim())) problems.push(`${name}: 스타일 속성과 기대값을 입력하세요.`);
      if (step.action === 'press' && !KEYS.includes(String(step.value))) problems.push(`${name}: 지원하는 키를 선택하세요.`);
    }
    if (rule.source.kind === 'explicit' && (!rule.source.quote?.trim() || !contract.assetIds.includes(rule.source.assetId || ''))) {
      problems.push(`${name}: 선택한 파일과 원문 근거를 확인하세요.`);
    }
  }
  return [...new Set(problems)];
}
export const roundApprovable = (round?: Round) => round?.passed === true && /^[a-f0-9]{64}$/i.test(round.artifactSha256) &&
  round.hasHtml === true && round.hasReport === true && round.hasScreenshot === true &&
  Array.isArray(round.blockingFindings) && round.blockingFindings.length === 0;
export function ocrLabel(status?: string): string {
  return ({ complete: '문자 인식 완료', partial: '일부만 문자 인식', failed: '문자 인식 실패',
    unavailable: '문자 인식 사용 불가', pending: '문자 인식 대기 중' } as Record<string, string>)[status || ''] || '문자 인식 미확인';
}
export function editable(contract: EditableContract): EditableContract {
  return { schemaVersion: 1, title: contract.title, brief: contract.brief, assetIds: [...contract.assetIds],
    ...Object.fromEntries(['projectId', 'productId', 'guidelineId', 'guidelineAssetId', 'ontologyHash', 'catalogHash']
      .filter(key => contract[key as keyof EditableContract] !== undefined).map(key => [key, contract[key as keyof EditableContract]])),
    viewport: { ...contract.viewport }, rules: structuredClone(contract.rules), unresolved: [...(contract.unresolved || [])],
    ...(contract.bindings !== undefined ? { bindings: { ...contract.bindings } } : {}) };
}
export const manualAssets = (assets: Asset[]) => assets.filter(asset => !asset.system && !asset.archived);
export function applyManualAssets(draft: EditableContract, assets: Asset[], selected: string[]): EditableContract {
  const allowed = new Set(manualAssets(assets).map(asset => asset.id));
  const fixed = draft.assetIds.filter(id => id === draft.guidelineAssetId || assets.some(asset => asset.id === id && asset.system));
  if (draft.guidelineAssetId) fixed.push(draft.guidelineAssetId);
  return { ...draft, assetIds: [...new Set([...selected.filter(id => allowed.has(id)), ...fixed])] };
}
export function importedHtmlAssets(assets: Asset[], contract?: Pick<EditableContract, 'assetIds'> | null): Asset[] {
  return assets.filter(asset => contract?.assetIds.includes(asset.id) && asset.uploadStatus === 'stored' &&
    !asset.system && !asset.archived && /\.html?$/i.test(asset.name));
}
export const isOriginalHtmlCheck = (run: Pick<Run, 'mode' | 'model'>) => run.mode === 'verify' || run.model === 'no-inference';
export function blankContract(assetIds: string[] = [], brief = ''): EditableContract {
  return { schemaVersion: 1, title: '새 화면의 확인 기준', brief, assetIds, viewport: { width: 390, height: 844 }, rules: [newRule()], unresolved: [] };
}
export function stateLabel(state?: string): string {
  return ({ queued: '대기 중', running: '진행 중', completed: '완료', needs_changes: '수정 필요', failed: '실패',
    'review-required': '화면 변형 검토 필요', committed: '커밋 완료',
    ready: '준비 완료', rebuilding: '재빌드 중', exporting: '내보내는 중', unavailable: '사용할 수 없음', needs_revalidation: '재검증 필요',
    passed: '통과', pass: '통과', fail: '실패', verified: '검증됨', not_run: '미검증', not_tested: '미검증',
    not_verified: '미검증', unverified: '미검증', pending: '미검증', incomplete: '미판정',
    'not-run': '미검증', not_compared: '비교하지 않음', skipped: '검사 생략', unsupported: '미지원', approved: '승인됨', draft: '작성 중',
    uploading: '전송 중', processing: '파일 확인 중', stored: '보관 완료', complete: '해석 완료', partial: '일부 해석',
  } as Record<string, string>)[state || ''] || '미확인';
}
